File: javascript/exercises-1-1.js
// Explanatory solutions are in the scheme directory

// Exercise 1.3

var sumOfSquaresOfTwoLargest = function (a, b, c) {
  var square = function (x) { return x * x; };
  var sumOfSquares = function (x, y) { return square(x) + square(y); };

  return sumOfSquares.apply(null, [a, b, c].sort().slice(1));
};


// utility functions for Exercise 1.7 & 1.8

var abs = function (x) {
  if (x < 0) return -x;
  return x;
};

var average = function (x, y) {
  return (x + y) / 2;
};


// Exercise 1.7

var sqrt = function (x) {
  var isGoodEnough = function (guess, lastGuess) {
    return abs(guess - lastGuess) / guess < 0.0000000001;
  };

  var improve = function (guess) {
    return average(guess, (x / guess));
  };

  var sqrtIter = function (guess, lastGuess) {
    if (isGoodEnough(guess, lastGuess))
      return guess;
    else
      return sqrtIter(improve(guess), guess);
  };
  
  return sqrtIter(1.0, 0.0);
};

// Exercise 1.8

var cubeRoot = function (x) {
  var square = function (x) { return x * x; };
  var cube = function (x) { return x * x * x; };

  var isGoodEnough = function (guess) {
    return abs(cube(guess) - x) < 0.0000000001;
  };
  
  var improve = function (guess) {
    return ((x / square(guess)) + (guess * 2)) / 3;
  };

  var cubeRootIter = function (guess) {
    if (isGoodEnough(guess))
      return guess;
    else
      return cubeRootIter(improve(guess));
  };

  return cubeRootIter(1.0);
};

File: clojure/exercises-1-1.clj
;; Skipping explanatory exercises.
;;  These responses are under the scheme directory.

; Exercise 1.3

(defn sum-of-squares-of-two-largest [a b c]
  (defn square [x] (* x x))
  (defn sum-of-squares [x y] (+ (square x) (square y)))
  (apply sum-of-squares
	 (drop 1 (sort (list a b c)))))


; Exercise 1.7

(defn sqrt [x]
  (defn abs [x] (if (< x 0) (- x) x))
  (defn average [a b] (/ (+ a b) 2))

  (defn good-enough? [guess last-guess]
    (< (/ (abs (- guess last-guess)) guess) 0.0000000001))

  (defn improve [guess]
    (average guess (/ x guess)))

  (defn sqrt-iter [guess last-guess]
    (if (good-enough? guess last-guess)
      guess
      (sqrt-iter (improve guess)
		 guess)))

  (sqrt-iter 1.0 0.0))


; Exercise 1.8


(defn cube-root [x]
  (defn square [x] (* x x))
  (defn cube [x] (* x x x))
  (defn abs [x] (if (< x 0) (- x) x))
  (defn average [x y] (/ (+ x y) 2))

  (defn improve [guess]
    (/ (+ (/ x (square guess)) (* 2 guess)) 3))

  (defn good-enough? [guess]
    (< (abs (- (cube guess) x)) 0.0000000001))

  (defn cube-root-iter [guess]
    (if (good-enough? guess)
        guess
        (cube-root-iter (improve guess))))

  (cube-root-iter 1.0))
